### ISSUE TITLE
Tilemap: Add 'bounce' or 'Ping-Pong' animation order to animated tiles

### ISSUE THREAD

**Issue by Proggle** [godotengine/godot-proposals#7643]
### Describe the project you are working on

Working on a game using a tile map with animated tiles.

### Describe the problem or limitation you are having in your project

There are a lot of animated tiles that go back and forth through animation frames instead of looping them in circle (like sloshing water, a clock's pendulum, etc).  

For a 'bounce' animation, I mean that instead of the animation following this 'linear' or 'loop' frame order:
frame 1
frame 2
frame 3
<loop back to frame 1>

It goes in this 'bounce' or 'ping-pong' sequence:
frame 1
frame 2
frame 3
frame 2
frame 1

As is, for a lot of these items I'd have to basically double the number of space these tiles took up in the sheet by duplicating them and putting the middle frames in twice in reverse order, which is wasteful.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be nice if there was a boolean under the 'animation' properties for a tile that I could click on which set it so that when the tile animation reached the end, it would reverse the frame order and go backwards right to the start.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

On the user end of it, it would be a boolean added under the 'animation' tab.

On the code end, you'd need to change 
 TileSetAtlasSource::get_tile_animation_total_duration to check for the boolean (to add the duration for middle tiles twice), and stick a second for loop at the end of TileMap::draw_tile to send the reverse-ordered tiles to the RenderingServer

### If this enhancement will not be used often, can it be worked around with a few lines of script?

In theory you can unwrap bounce animations into loop animations by basically doubling the amount of frames you use on your tilesheet, but that's more work, creates a bunch of duplicate tiles, and means that godot's animation system isn't compatible with existing tilesets that assume you can bounce your animations.

### Is there a reason why this should be core and not an add-on in the asset library?

This is a very common animation functionality in games (to the point where a lot of RPGs only have 3-frame 'bounce' animations in their spritesheets)

**Comment by Proggle** [comment]
Have implemented it in this PR https://github.com/godotengine/godot/pull/82232

**Comment by kleonc** [comment]
The feature makes sense to me.

My only concern is whether the last/first frames should be repeated (the same for #11698). For a time-based ping-pong (like in AnimationPlayer) both first and last frame should be repeated (played once for both going back and forth).
```
1-2-3-4-1-2-3-4-1-2-3-4-1-2-3-4-... // Normal looping.
^       ^       ^       ^      

1-2-3-4-3-2-1-2-3-4-3-2-1-2-3-4-... // Ping-pong proposed here (cycle lasts `2 * anim_duration - (first_frame_duration + last_frame_duration)`).
^     ^     ^     ^     ^     ^
1-2-3-4-4-3-2-1-1-2-3-4-4-3-2-1-... // Time-based ping-pong (cycle lasts `2 * anim_duration`).
^       ^       ^       ^      
```

**Comment by TokageItLab** [comment]
```
1-2-3-4-3-2-1-2-3-4-3-2-1-2-3-4-... // Ping-pong proposed here (cycle lasts `2 * anim_duration - (first_frame_duration + last_frame_duration)`).
^     ^     ^     ^     ^     ^
```

I think support for the scenario described above is generally sufficient.

Since each frame has a **timescale**, I assume you can achieve the goal by doubling the duration of the end frames. The only concern is that this method would double the duration of the first frame, like:

```
a 1: x2
b 2: x1
c 3: x1
d 4: x2

1--2-3-4--3-2-1--2-3-4--3-2-1--...
```

so we should document that `frame_progress` should be set to `0.5` when playing such time-based pingpong animation.

Or we might able to setting below if it can handle 0 length frame:

```
a 1: x0
a 2: x1
b 3: x1
c 4: x1
d 5: x1
d 6: x0
``` 

In any case, as long as we have Case 1234321234, there is a workaround for 123443211234. Conversely, case 123443211234 is not 
including with the 1234321234 case (so 123443211234 cases ⊂ 1234321234 cases).

**Comment by Proggle** [comment]
As mentioned in the sprite2d pingpong proposal, I think a lot of users would be surprised if frame 1 was listed as lasting 1 second and instead lasted for 2 seconds on the screen.
Having the listed durations be accurate makes thr most sense, if you want frame one to be twice as long you can set the duration appropriately.

**Comment by kleonc** [comment]
> Since each frame has a **timescale**, I assume you can achieve the goal by doubling the duration of the end frames. The only concern is that this method would double the duration of the first frame

Indeed, doubling the first/last frame durations would make the ping-pong cycle take `2 * anim_duration`, but there would be phase shift at the beginning (looped and ping-pong animations with the same frames started at the same time would no be synced). So such workaround would not be a one-to-one substitute .

> Or we might able to setting below if it can handle 0 length frame:

`0.0` length frames would indeed do as a one-to-one workaround. But note that currently we don't support 0 length (talking about #11698 at the same time as I think they should be implemented in the same manner).

https://github.com/godotengine/godot/blob/f9a8302c8395db323fdeebcb1d8024b9fc661999/scene/resources/2d/tile_set.cpp#L5156
https://github.com/godotengine/godot/blob/f9a8302c8395db323fdeebcb1d8024b9fc661999/scene/resources/sprite_frames.cpp#L61
https://github.com/godotengine/godot/blob/f9a8302c8395db323fdeebcb1d8024b9fc661999/scene/resources/sprite_frames.h#L35

---

Nevertheless, I agree it's fine to go with the frame-based ping-pong proposed here. If there will be a demand for the time-based ping-pong then it could possibly be added later (either by adding separate `PING_PING_TIMEWISE` (or alike) mode, or by allowing `0.0` frame durations to allow a workaround). For now frame-based ping-pong only should be fine (it just should be clearly documented that it's frame based). 👍
